src/firmware/types.rs

```diff
@@ -202,3 +210,512 @@ pub struct SnpPlatformStatus {
+    }
+
+    /// Bit 18 of Guest Policy's info
+    pub fn single_socket(&self) -> bool {
+        bit_val!(self.info, 18, u64) != 0
+    }
+
+    /// Bit 19 of Guest Policy's info
+    pub fn debug(&self) -> bool {
+        bit_val!(self.info, 19, u64) != 0
+    }
+
+    /// Bit 20 of Guest Policy's info
+    pub fn migrate_ma(&self) -> bool {
+        bit_val!(self.info, 20, u64) != 0
+    }
+}
+
+/// The TCB_VERSION is a structure containing the security version numbers of each component in
+/// the trusted computing base (TCB) of the SNP firmware. associated A TCB_VERSION is
+/// associated with each image of firmware. T
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+#[cfg_attr(feature = "openssl", derive(Serialize, Deserialize))]
+
+pub struct SNPTcbVersion {
+    /// Current bootloader version. SVN of PSP Bootloader.
+    pub boot_loader: u8,
+    /// Current PSP OS version. SVN of PSP Operating System.
+    pub tee: u8,
+    reserved: [u8; 4],
+    /// Version of the SNP firmware. Security Version Number (SVN) of SNP firmware.
+    pub snp: u8,
+    /// Lowest current patch level of all the cores.
+    pub microcode: u8,
+}
+
+impl SNPTcbVersion {
+    pub fn raw(&self) -> String {
+        format!(
+            "{:02}{:02}{:02}{:02}{:02}{:02}{:02}{:02}",
+            self.boot_loader,
+            self.tee,
+            self.reserved[3],
+            self.reserved[2],
+            self.reserved[1],
+            self.reserved[0],
+            self.snp,
+            self.microcode,
+        )
+    }
+}
+
+/// A structure with a bit-field unsigned 64 bit integer:
+/// Bit 0 representing the status of TSME enablement.
+/// Bit 1 representing the status of SMT enablement.
+/// Bits 2-63 are reserved.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default)]
+#[cfg_attr(feature = "openssl", derive(Serialize, Deserialize))]
+pub struct PlatformInfo {
+    info: u64,
+}
+
+impl PlatformInfo {
+    /// Returns the state of TSME
+    pub fn tsme_en(&self) -> bool {
+        bit_val!(self.info, 1, u64) != 0
+    }
+
+    /// Returns the state of SMT
+    pub fn smt_en(&self) -> bool {
+        bit_val!(self.info, 2, u64) != 0
+    }
+}
+
+/// The format of an ECDSA P-384 with SHA-384 Signature.
+#[repr(C)]
+#[derive(Debug)]
+#[cfg_attr(feature = "openssl", derive(Serialize, Deserialize))]
+pub struct Signature {
+    /// R component of this signature. Value is zero-extended little-endian encoded.
+    #[cfg_attr(feature = "openssl", serde(with = "BigArray"))]
+    pub r: [u8; 72],
+    /// S component of this signature. Value is zero-extended little-endian encoded.
+    #[cfg_attr(feature = "openssl", serde(with = "BigArray"))]
+    pub s: [u8; 72],
+    #[cfg_attr(feature = "openssl", serde(with = "BigArray"))]
+    _reserved_0: [u8; 512 - 144],
+}
+
+#[cfg(feature = "openssl")]
+impl Signature {
+    pub fn get_ecdsa_sig(&mut self) -> EcdsaSig {
+        // OpenSSL Expects BigNum values in Big Endian so use lebin2bn
+
```

Comment:
1. why `mut self` here?
2. consider returning `Result<EcdsaSig>` instead of panic! inside.

Reply:
That is definitely where I would like to take this (hadn't gotten that far yet). Do you think it would be worth making a set of errors for this, or just pass up the BigNum C-errors instead?